core/src/test/java/io/temporal/samples/earlyreturn/TransactionWorkflowTest.java

```diff
@@ -0,0 +1,159 @@
+            new TransactionRequest(SOURCE_ACCOUNT, TARGET_ACCOUNT, VALID_AMOUNT),
+            updateOp);
+
+    // Verify both update and final results
+    TxResult updateResult = handle.getResultAsync().get();
+    assertEquals(TEST_TRANSACTION_ID, updateResult.getTransactionId());
+
+    TxResult finalResult = WorkflowStub.fromTyped(workflow).getResult(TxResult.class);
+    assertEquals("Transaction completed successfully.", finalResult.getStatus());
+
+    // Verify activities were called
+    verify(activities).initTransaction(any());
+    verify(activities).completeTransaction(any());
+    verifyNoMoreInteractions(activities);
+  }
+
+  @Test
+  public void testUpdateWithStartInvalidAmount() throws Exception {
+    // Mock activities
+    TransactionActivities activities =
+        mock(TransactionActivities.class, withSettings().withoutAnnotations());
+    when(activities.initTransaction(any()))
+        .thenThrow(
+            ApplicationFailure.newNonRetryableFailure(
+                "Non-retryable Activity Failure: Invalid Amount", "InvalidAmount"));
+    doNothing().when(activities).cancelTransaction(any());
+
+    testWorkflowRule.getWorker().registerActivitiesImplementations(activities);
+    testWorkflowRule.getTestEnvironment().start();
+
+    // Create workflow stub with explicit ID
+    WorkflowClient workflowClient = testWorkflowRule.getWorkflowClient();
+    String workflowId = "test-workflow-" + UUID.randomUUID();
+    WorkflowOptions options =
+        WorkflowOptions.newBuilder()
+            .setTaskQueue(testWorkflowRule.getTaskQueue())
+            .setWorkflowId(workflowId)
+            .build();
+
+    TransactionWorkflow workflow =
+        workflowClient.newWorkflowStub(TransactionWorkflow.class, options);
+
+    // Create update operation
+    UpdateWithStartWorkflowOperation<TxResult> updateOp =
+        UpdateWithStartWorkflowOperation.newBuilder(workflow::returnInitResult)
+            .setWaitForStage(WorkflowUpdateStage.COMPLETED)
+            .build();
+
+    // Execute UpdateWithStart and expect the exception
+    WorkflowServiceException exception =
+        assertThrows(
+            WorkflowServiceException.class,
+            () ->
+                WorkflowClient.updateWithStart(
+                    workflow::processTransaction,
+                    new TransactionRequest(SOURCE_ACCOUNT, TARGET_ACCOUNT, INVALID_AMOUNT),
+                    updateOp));
+
+    // Verify the exception chain
+    assertTrue(exception.getCause() instanceof WorkflowUpdateException);
+    assertTrue(exception.getCause().getCause() instanceof ActivityFailure);
+    ApplicationFailure appFailure = (ApplicationFailure) exception.getCause().getCause().getCause();
+    assertEquals("InvalidAmount", appFailure.getType());
+    assertTrue(appFailure.getMessage().contains("Invalid Amount"));
+
+    // Let workflow process error handling
+    testWorkflowRule.getTestEnvironment().sleep(Duration.ofSeconds(1));
```

Comment:
Out of curiosity, is this sleep necessary? I would have expected the `getResult` further down would just block until the result is available?

Reply:
You're absolutely right. There's no need as getResult blocks for a response. Removing